src/lib/server/hyrise_session.cpp

```diff
@@ -23,278 +23,270 @@
-  }
-
-  // We're currently waiting for a query, so accept the incoming one
-  switch (_input_packet_type) {
-    case NetworkMessageType::SimpleQueryCommand:
-      return _accept_query();
-    case NetworkMessageType::ParseCommand:
-      return _accept_parse();
-    case NetworkMessageType::BindCommand:
-      return _accept_bind();
-    case NetworkMessageType::DescribeCommand:
-      return _accept_describe();
-    case NetworkMessageType::SyncCommand:
-      return _accept_sync();
-    case NetworkMessageType::FlushCommand:
-      return _accept_flush();
-    case NetworkMessageType::ExecuteCommand:
-      return _accept_execute();
-    default: {
-      _send_error("Unsupported message type");
-      _send_ready_for_query();
+boost::future<void> HyriseSession::_handle_client_requests() {
+  auto process_request_header = [=](RequestHeader request) {
+    // Process the request header
+
+    if (request.message_type == NetworkMessageType::TerminateCommand)
+      // Not really unexpected, but this is a way to escape the infinite recursion
+      throw std::logic_error("Session was terminated by client");
+
+    // Proceed by reading the packet contents
+    return _connection->receive_packet_contents(request.payload_length)
+      >> then >> [=] (InputPacket packet_contents) {
+      return std::make_pair(request, std::move(packet_contents)); 
+    };
+  };
+  
+  auto process_command = [=](std::pair<RequestHeader, InputPacket> packet) {
+    switch (packet.first.message_type) {
+      case NetworkMessageType::SimpleQueryCommand: {
+        const auto sql = PostgresWireHandler::handle_query_packet(packet.second);
+        return _handle_simple_query_command(sql)
+          >> then >> [=]() { _connection->send_ready_for_query(); };
+      }
+      case NetworkMessageType::ParseCommand:
+        // TODO
+        // return _accept_parse();
+        return boost::make_ready_future();
+
+      case NetworkMessageType::BindCommand:
+        // TODO
+        // return _accept_bind();
+        return boost::make_ready_future();
+
+      case NetworkMessageType::DescribeCommand:
+        // TODO
+        // return _accept_describe();
+        return boost::make_ready_future();
+
+      case NetworkMessageType::SyncCommand:
+        // TODO
+        // return _accept_sync();
+        return boost::make_ready_future();
+
+      case NetworkMessageType::FlushCommand:
+        // TODO
+        // return _accept_flush();
+        return boost::make_ready_future();
+
+      case NetworkMessageType::ExecuteCommand:
+        // TODO
+        // return _accept_execute();
+        return boost::make_ready_future();
+
+      default:
+        throw std::logic_error("Unsupported message type");
+    }
+  };
+  
+  auto command_result = 
+    _connection->receive_packet_header() 
+      >> then >> process_request_header
+      >> then >> process_command;
+  
+  return command_result.then(boost::launch::sync, [=] (boost::future<void> result) {
```

Comment:
can we split this up somehow? hard to tell in which case what happens with nested `>> then >>` and then the unwrap and then again...